impl/src/testing.rs

```diff
@@ -56,3 +96,107 @@ pub fn make_workspace(
 pub fn make_basic_workspace() -> (TempDir, CargoWorkspaceFiles) {
   make_workspace(basic_toml(), Some(basic_lock()))
 }
+
+/** Creates mock responses for fetching crate metadata and source code from a crate registry */
+pub fn mock_remote_crate(name: &str, version: &str, mock_server: &MockServer) -> TempDir {
+  // Crate info mock response
+  mock_server.mock(|when, then| {
+    when.method(GET).path(format!("/api/v1/crates/{}", name));
+    then.status(200).json_body(json!({
+        "crate": {
+            "id": name,
+            "name": name,
+            "versions": [
+                283716
+            ],
+        },
+        "versions": [
+            {
+                "id": 283716,
+                "crate": name,
+                "num": version,
+                "dl_path": format!("/api/v1/crates/{}/{}/download", name, version),
+            }
+        ],
+    }));
+  });
+
+  // Create archive contents
+  let dir = tempfile::TempDir::new().unwrap();
+  let tar_path = dir.as_ref().join(format!("{}.tar.gz", name));
+  {
+    std::fs::create_dir_all(dir.as_ref().join("archive")).unwrap();
+    std::fs::File::create(dir.as_ref().join("archive/test")).unwrap();
+    std::fs::write(dir.as_ref().join("archive/Cargo.toml"), named_toml(name, version)).unwrap();
+    std::fs::write(dir.as_ref().join("archive/Cargo.lock"), named_lock(name, version)).unwrap();
+
+    let tar_gz: std::fs::File = std::fs::File::create(&tar_path).unwrap();
+    let enc = flate2::write::GzEncoder::new(tar_gz, Compression::default());
+    let mut tar = tar::Builder::new(enc);
+    tar
+      .append_dir_all(
+        format!("{}-{}", name, version),
+        dir.as_ref().join("archive"),
+      )
+      .unwrap();
+  }
+
+  // Create download mock response
+  mock_server.mock(|when, then| {
+    when
+      .method(GET)
+      .path(format!("/api/v1/crates/{}/{}/download", name, version));
+    then
+      .status(200)
+      .header("content-type", "application/x-tar")
+      .body_from_file(&tar_path.display().to_string());
+  });
+
+  dir
+}
+
+/** A helper macro for passing a `crates` to  `mock_crate_index` */
+#[macro_export]
+macro_rules! index_crates {
```

Comment:
This can be a function, right?

(again, use the _least powerful_ available tool for the job)